src/libs/ajax/loaded-data/useLoadedData.ts

```diff
@@ -50,11 +64,15 @@ export const useLoadedData = <T>(): UseLoadedDataResult<T> => {
       const error = isFetchResponse(err) ?
         Error(await err.text()) :
         err
-      setLoadedData({
+      const errorResult: ErrorState<T, unknown> = {
         status: 'Error',
         state: previousState,
         error
-      })
+      }
+      setLoadedData(errorResult)
+      if (args.onError) {
+        args.onError(errorResult)
+      }
```

Comment:
Can this be written as
```suggestion
      args?.onError(errorResult)
```

Reply:
it could (sortof)... but I find folks get confused and/or over/misuse this operator... it only validates the LEFT item in the chain (args) in this case to not be null or undefined)... in this case args is already guaranteed to not be null because the type safety on line 48 above this where args is declared.  What I think you are suggesting is this:
`args.onError.?(errorResult)`
as detailed in the "optional call"  paragraph of https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining

I can give this a try... as long as the instrumentation of proper unit test branch coverage check out, I'll get it in before merge.

Reply:
uh oh.... tried:
`args.onError?.(errorResult)`
and the branch coverage falsely reported 100% without me testing the args.onError == null or undefined case.

Looks like I will need to add a note to advise caution/should-avoid on this since it currently cripple code coverage checks for important cases that should be tested.

Reply:
Yes, `args.onError?.(errorResult)` is what I meant to type

Reply:
> and the branch coverage falsely reported 100% without me testing the args.onError == null or undefined case.

IMO this is fine and I wouldn't expect to have to write a test to validate the `?.` is working properly. That's the benefit of using a language builtin vs a `? :` or `||` operation to handle undefined state.

Reply:
the danger is that these are syntactic sugar for code equivalents that are code branches.   In this example, not being alerted of the branch coverage gap means that there is a good chance that a test with good assertions around the onError callback correctly being called if provided never gets written.  Code changes could break this expectation and there would be no signal that consumers' expectations just got broken.

Reply:
I understand that typescript converts its code to javascript so its "primitives" are really implemented as javascript code. And in this case, yes, `x?.y` is converted to a conditional (I think it becomes `x && x.y`).

What I'm saying is that, unlike `x && x.y`, there's no need to test the case where `x` is false, because we know that typescript generates the correct code. As long as the positive case has a test, that should be sufficient. If you're saying that the coverage report doesn't require testing the _positive_ case, then I agree, it's a bug.

Reply:
right... and that is the danger of some of these convenience patterns... they easily hide code branches, in sometimes significantly negative ways.  Let's define code branches as "the code can produce more then one outcome", whether thru more conventional if statements, language conveniences, or anything in between. The underlying mechanic doesn't matter... just that it produces more then one possible outcome in the code it's being used in.   In the example above, as long as one test hits the line args.onError.?(errorResult) with the falsy case, it's considered completely covered... hence creating the positive case not needing to be tested scenario.

The Typescript to Javascript flow is it's own beast... First you have the conversion to be careful of (usually babel.js under the hood), and then there is the intrinsically dynamic elements of JavaScript itself.  I am curious to look into possible tooling solutions out there to avoid the branch hiding of some of these... but that's another track.

Reply:
I ran some tests and it looks like the jest coverage data does take both branches into consideration when generating branch coverage data for `x?.y()`. Here's what I did. The first run doesn't use `onError`, the second does, and the final run executes all loaded data tests. Note that the branch coverage is the second column.
```
$ yarn run --binaries-only react-scripts test --coverage --testNamePattern='^useLoadedData hook handles error state from Fetch Response as error$' --runTestsByPath src/libs/ajax/loaded-data/useLoadedData.test.ts | grep useLoadedData.ts
useLoadedData.ts                                 |    92.3 |       60 |     100 |    92.3 | 61
$ yarn run --binaries-only react-scripts test --coverage --testNamePattern='^useLoadedData hook handles ready state, then later error state$' --runTestsByPath src/libs/ajax/loaded-data/useLoadedData.test.ts | grep useLoadedData.ts
useLoadedData.ts                                 |     100 |       60 |     100 |     100 | 50,66
$ yarn run --binaries-only react-scripts test --coverage --testNamePattern='^useLoadedData hook'  --runTestsByPath src/libs/ajax/loaded-data/useLoadedData.test.ts | grep useLoadedData.ts
useLoadedData.ts                                 |     100 |      100 |     100 |     100 |
$
```
Based on this, it looks like there's no harm in using `?.` with respect to code coverage.

Reply:
created a UIE ticket to track the exploration of a tooling fix:
https://broadworkbench.atlassian.net/browse/UIE-68